### PR TITLE
Mac keycode refactor

### PR DESCRIPTION
**Pull request by MuhammedZakir**
Ref: https://github.com/kmonad/kmonad/pull/259
Resolves: https://github.com/kmonad/kmonad/issues/330

### CODE REVIEW

**Comment by MuhammedZakir** [comment]
For this to work, comment out `cmps` and `102d` keycode aliases in `src/KMonad/App/Parser/Keycode.hs`. Non-common aliases should be moved to OS-specific files?

File: src/KMonad/App/KeyIO/Mac/IOKitSource.hs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        case (prodStr) of
+          Nothing -> do
+            logInfo "Opening HID devices"
+            liftIO . void $ grab_kb nullPtr
+          Just s -> do
+            logInfo $ "Opening HID device: " <> s
+            liftIO $ do
+              str <- newCString $ unpack s
+              void $ grab_kb str
+              free str
```
Comment (slotThe) [inline src/KMonad/App/KeyIO/Mac/IOKitSource.hs:56]: Everytime you have this `newCString` -> do stuff -> `free` pattern going on its worth looking at whether `withCString` may be a better fit

File: src/KMonad/App/KeyIO/Mac/Kext.hs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  send_key :: Ptr RawEvent -> OnlyIO ()
+
+-------------------------------------------------------------------------------
+
+withKext :: LUIO m e => KextCfg -> Ctx r m PutKey
+withKext _ = mkCtx $ \f -> do
+
+  let init = do
+        logInfo "Initializing Mac key sink"
+        liftIO $ mallocBytes $ sizeOf (undefined :: RawEvent)
```
Comment (slotThe) [inline src/KMonad/App/KeyIO/Mac/Kext.hs:30]: Same comment applies here with regards to `undefined` vs. `Proxy`

File: src/KMonad/Util/Keyboard/Mac.hs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+-- https://stackoverflow.com/questions/1959715/how-to-print-integer-literals-in-binary-or-hex-in-haskell/55623656#55623656
+toHex :: Word32 -> String
+toHex = ("0x" ++) . convertToBase 16
+  where
+    convertToBase :: Word8 -> Word32 -> String
+    convertToBase b n
+      | n < 0              = '-' : convertToBase b (-n)
+      | n < fromIntegral b = [(['0'..'9'] ++ ['A' .. 'Z']) !! fromIntegral n]
+      | otherwise          = let (d, m) = n `divMod` fromIntegral b in convertToBase b d ++ convertToBase b m
```
Comment (slotThe) [inline src/KMonad/Util/Keyboard/Mac.hs:51]: What's the reasoning for prefering this to `Numeric.showHex`?

File: src/KMonad/Util/Keyboard/Mac.hs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+mkRaw p c = RawEvent
+  { _reCode = unKeycode c
+  , _reVal  = p ^. re _EvType }
+
+
+--------------------------------------------------------------------------------
+-- $keynames
+
+-- | zip 2 lists only if they have the same length.
+--   Caution: only ever use this to generate static data.
```
Comment (slotThe) [inline src/KMonad/Util/Keyboard/Mac.hs:109]: I think this is a bit misleading, as it suggests that the only thing that's happening is a length comparison and then a `zip`

File: src/KMonad/Util/Keyboard/Mac.hs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+z :: Word32 -> [Word32] -> [Text] -> [(CoreName, Keycode)]
+z page codes names
+  | length codes == length names = zipWith func codes names
+  | otherwise            = error "mac: keycodes length"
+  where
+    func = (\c n -> (CoreName n, Keycode (page, c)))
```
Comment (slotThe) [inline src/KMonad/Util/Keyboard/Mac.hs:115]:
```suggestion
    func c n = (CoreName n, Keycode (page, c))
```

File: src/KMonad/Util/Keyboard/Mac.hs [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+  , z 0xFF01 [0x20..0x21] [ "Vbru", "Vbrd" ]
+
+  -- ###  0xFF => AppleVendor Page Top Case  ###
+  , z 0xFF [0x3..0x5] [ "Fn", "brup", "brdn" ]
+  -- kbil -> IlluminationToggle
+  , z 0xFF [0x7..0x9] [ "kbil", "kbup", "kbdn" ]
+  ]
+   where
+      -- Apple uses "1-9 & 0" order for numbers
+      num_1To9 = map tshow $ [1..9 :: Int]
```
Comment (slotThe) [inline src/KMonad/Util/Keyboard/Mac.hs:184]:
```suggestion
      num_1To9 = map tshow [1..9 :: Int]
```

File: src/KMonad/Util/Keyboard/Mac.hs [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+toHex :: Word32 -> String
+toHex = ("0x" ++) . convertToBase 16
+  where
+    convertToBase :: Word8 -> Word32 -> String
+    convertToBase b n
+      | n < 0              = '-' : convertToBase b (-n)
+      | n < fromIntegral b = [(['0'..'9'] ++ ['A' .. 'Z']) !! fromIntegral n]
+      | otherwise          = let (d, m) = n `divMod` fromIntegral b in convertToBase b d ++ convertToBase b m
+--------------------------------------------------------------------------------
+  -- $evtype
```
Comment (slotThe) [inline src/KMonad/Util/Keyboard/Mac.hs:53]:
```suggestion
-- $evtype
```

Otherwise Haddock is confused (at least on my system)

**Review by slotThe — COMMENTED** [review]
Thank you for taking the initiative!  Due to not having a Mac, I can only review the code as-is (that is, without testing it).  

Cc. @david-janssen for a possible inclusion in #259

File: src/KMonad/App/KeyIO/Common.hs [hunk 8]
```diff
@@ -50,12 +50,17 @@ instance Default UinputCfg where
 --------------------------------------------------------------------------------
 -- $cfgs-mac
 
--- TODO: This is where the Mac config records go
+-- | Holds the product string of keyboard if given. Else, Nothing.
+data IOKitCfg = IOKitCfg
```
[the remark this answers is not in the document]
Comment (MuhammedZakir) [reply src/KMonad/App/KeyIO/Common.hs:54]: Should I also change `EvdevCfg`?

File: src/KMonad/Util/Keyboard/Mac.hs (hunk 4 above)
Comment (MuhammedZakir) [reply src/KMonad/Util/Keyboard/Mac.hs:109]: What would be a better description?

File: src/KMonad/App/KeyIO/Common.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (slotThe) [reply src/KMonad/App/KeyIO/Common.hs:54]: I would say so, yes

File: src/KMonad/Util/Keyboard/Mac.hs (hunk 4 above)
Comment (slotThe) [reply src/KMonad/Util/Keyboard/Mac.hs:109]: Good question :)  Since this is not polymorphic, I guess we could write exactly what's going on; something about building up named keycode pairs, perhaps

**Comment by MuhammedZakir** [comment]
Two questions:

1.  > Apple internal keyboard wraps key events with (iirc) `(0x7, 0xFF)` keycode. For example, when tapping `a`, instead of `Pa Ra`, we get `P(0x7, 0xFF) Pa Ra R(0x7, 0xFF)`. IIAR, we must ignore this keycode when capturing events, but I am an not sure whether we should emit this when sending events. FWIW, I used Karabiner-EventViewer for testing, and I don't get this when I use Karabiner-Elements.
    > 
    > tl;dr -- should we wrap key [send] events with key event of this keycode, or should we ignore it?

2. Although we support both Kext and Dext and the config for both is the same, the word "Kext" is used in many of the names of types and functions. I think it is better to use a general name for clarity, but I can't think of a name. Any suggestion?

File: src/KMonad/App/KeyIO/Mac/IOKitSource.hs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+        liftIO . mallocBytes $ sizeOf (undefined :: RawEvent)
```
[the remark this answers is not in the document]
Comment (MuhammedZakir) [reply src/KMonad/App/KeyIO/Mac/IOKitSource.hs:58]: Didn't know about that. Thanks! :-)

File: src/KMonad/Util/Keyboard/Mac.hs (hunk 4 above)
Comment (MuhammedZakir) [reply src/KMonad/Util/Keyboard/Mac.hs:109]:
```
-- | Takes a usage page, a list of usages and a list of names, 
--    and returns a list of tuples of keycodes and their names.
--
-- Caution: Only use this for static data as this throws an error
-- if the lengths of lists differ.
```

I am not familiar with Haddock; if there is any mistake in markup, feel free to correct me.

**Comment by MuhammedZakir** [comment]
In this amended commit, I have also changed `waitKey` to filter `(0x7, 0xFFFF)` keycode. I coverted this PR to draft as I have to do a few tests regarding this. Will change it back once that's done.

File: src/KMonad/Util/Keyboard/Mac.hs (hunk 4 above)
Comment (slotThe) [reply src/KMonad/Util/Keyboard/Mac.hs:106]: Looks good to me :+1:

**Comment by slotThe** [comment]
Sorry for not getting back to you sooner; I've been having lots of IRL things coming up with trying to finish my thesis. 

>     1. > Apple internal keyboard wraps key events with (iirc) `(0x7, 0xFF)` keycode. For example, when tapping `a`, instead of `Pa Ra`, we get `P(0x7, 0xFF) Pa Ra R(0x7, 0xFF)`. IIAR, we must ignore this keycode when capturing events, but I am an not sure whether we should emit this when sending events. FWIW, I used Karabiner-EventViewer for testing, and I don't get this when I use Karabiner-Elements.
>        > tl;dr -- should we wrap key [send] events with key event of this keycode, or should we ignore it?

Karabiner is the de facto standard remapping utility for Mac, right?  In this case I think emulating what they do can't be _that_ bad

>     2. Although we support both Kext and Dext and the config for both is the same, the word "Kext" is used in many of the names of types and functions. I think it is better to use a general name for clarity, but I can't think of a name. Any suggestion?

Are these shorthand for kernel extensions and driver extensions?  In this case something like like just extensions might fit the bill (but I'm bad with names so don't listen to me :))

**Comment by MuhammedZakir** [comment]
> Sorry for not getting back to you sooner; I've been having lots of IRL things coming up with trying to finish my thesis.

Np. Good luck with your thesis! :-)

> Karabiner is the de facto standard remapping utility for Mac, right? In this case I think emulating what they do can't be _that_ bad

So we will ignore that keycode.

> Are these shorthand for kernel extensions and driver extensions? In this case something like like just extensions might fit the bill (but I'm bad with names so don't listen to me :))

I will use `Ext`. Can change if someone gives a better name.

**Comment by MuhammedZakir** [comment]
I got some compilation errors after the latest change. I probably won't be able to fix them in coming days. I will post here after they are fixed.

**Comment by MuhammedZakir** [comment]
I think I have fixed the compilation errors except when I use `sizeOf (Proxy @RawEvent)` instead of `sizeOf (undefined :: RawEvent), I get

```
    • Could not deduce (Storable (Proxy RawEvent))
        arising from a use of ‘sizeOf’
      from the context: LUIO m e
        bound by the type signature for:
                   withIOKitSource :: forall (m :: * -> *) e r.
                                      LUIO m e =>
                                      IOKitCfg -> Ctx r m GetKey
        at /home/user/dev/haskell/kmonad/src/KMonad/App/KeyIO/Mac/IOKitSource.hs:59:1-57
    • In the second argument of ‘($)’, namely
        ‘sizeOf (Proxy @RawEvent)’
      In a stmt of a 'do' block:
        liftIO . mallocBytes $ sizeOf (Proxy @RawEvent)
      In the expression:
        do case prodStr of
             Nothing -> do ...
             Just s -> do ...
           liftIO . mallocBytes $ sizeOf (Proxy @RawEvent)
   |
74 |         liftIO . mallocBytes $ sizeOf (Proxy @RawEvent)
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by slotThe** [comment]
> I think I have fixed the compilation errors except when I use `sizeOf (Proxy @RawEvent)` instead of `sizeOf (undefined :: RawEvent)`, I get
> [...]

Whoops, sorry about that; I must have misread the type signature of `sizeOf`.  In this case, `Proxy` will not work and `undefined` is the way to go

**Comment by MuhammedZakir** [comment]
> Whoops, sorry about that; I must have misread the type signature of `sizeOf`. In this case, `Proxy` will not work and `undefined` is the way to go

Np.

---

Now the only thing left to do is to do some tests. It will take a few days. I will post here once that's done.

**Comment by nightscape** [comment]
I'm just testing this on [my BONE keyboard layout](https://github.com/nightscape/kmonad/blob/nightscape/keymap/user/nightscape/bone_mac.kbd) and it's throwing the following error:
```
kmonad: NonExistentTarget ("compose","cmps")
```
Surprising thing is that I didn't find anything in the diffs of this PR regarding `compose` or `cmps`...
@MuhammedZakir did you ever encounter this problem?

**Comment by MuhammedZakir** [comment]
@nightscape: It is because of aliases.

https://github.com/kmonad/kmonad/pull/346#issuecomment-903878980:
> For this to work, comment out keycode aliases for `cmps` and `102d` in `src/KMonad/App/Parser/Keycode.hs`. 

---

@slotThe @david-janssen :
> Aliases for key codes that are not in all 3 OSs should be moved to OS-specific files?

**Comment by david-janssen** [comment]
Hey there,

Sorry for my loooooong absenses, both in the past and probably in the future. Thank you very much for taking point on this, I wouldn't know where to start with Mac stuff.

> Edit: Simples solution is to move OS-specific aliases in Util/Keyboard/<os>.hs, and then merge them into keycodeAliases in App/Parser/Keycode.hs. I guess this could work. What do you think?

That seems like a good solution. It's been a few months since I dove into code, so it's all a bit hazy at the moment. Trying to get my bearings at the moment. IIRC, the plan was to have a single set of names for keys to use internally, to provide a collection of aliases inside the parser, and document very carefully which names should work cross-platform and which shouldn't.

The aliases should be set up in such a way that a standard 100% keyboard (ISO or ANSI) can be easily described for any OS.

@MuhammedZakir If there is anywhere you'd like me to look at first, let me know. I'm hoping to start dripping a little bit of time into KMonad so long as my health holds.

**Comment by david-janssen** [comment]
Huzzah! I managed (after great headaches) to actually get a MacOS virtualbox up and running, and even to install the `dext` and compile KMonad. I can run KMonad, and looking at the verbose logging, it correctly captures keyboard input and internally processes it well enough, but I can't get KMonad to produce keyboard output.

I have no idea if this is because I'm running inside a virtualbox, or if there is some other reason, and I wouldn't know where to begin with debugging this issue.

In regards to the aliases, the 2 broken ones are:
'compose' and '102d'. We do need to figure out how you call those in OSX, because they refer to buttons that exist on a standard 100% keyboard. 102nd is the key between `lshift` and `z` on ISO-keyboards. `compose` is the key between `alt gr` or `right-alt` and `rctrl`. We need to figure out some naming-scheme that is supported across all 3 platforms to address every single key on 100% keyboards.

I'll move on to working on the windows keycode refactor for now and wait for you to respond.

Cheerio

**Comment by MuhammedZakir** [comment]
Have you allowed `Accessibility` for the terminal/app you are using to run KMonad? Is your installed `Dext`'s version correct? Also, I don't remember, but I think you may have to run KMonad using sudo.

---

I don't have an ISO keyboard, so this is just my guess: I think the key between `lshift` and `z` is `(0x7, 0x64) - NonUS Backslash` and the `compose` key is probably `(0x7, 0x65) - Application`. If you (or someone else) have ISO keyboard, then you can get the keycode using Karabiner-EventViewer.

**Comment by david-janssen** [comment]
> I don't have an ISO keyboard, so this is just my guess: I think the key between lshift and z is (0x7, 0x64) - NonUS Backslash and the compose key is probably (0x7, 0x65) - Application. If you (or someone else) have ISO keyboard, then you can get the correct keycode using Karabiner-EventViewer.

Those both sound very plausible. I don't have either an ISO keyboard I can plug into anything, but maybe I can ask around on the Karabiner github or something.

> Have you allowed Accessibility for the terminal/app you are using to run KMonad? Is your installed Dext's version correct? Also, I don't remember, but I think you may have to run KMonad using sudo.

I needed to go through all of those steps just to get KMonad to run, and I followed the Mac installation instructions to the letter. I think, however, I'm basically just happy that I can see it compiles correctly without intending to tinker with Mac alot myself. It could very well be an issue with running in a virtualbox, since Mac already makes that very annoying, and I think, technically in breach of the user-agreement...

How should we proceed from here? Shall I have a go at merging this PR in or are there any remaining issues that you are aware of on the Mac side of things? After we merge this in I have some serious tinkering to do with the Windows stuff, that code isn't even finished yet, but I have had success in the past developing through VirtualBox.

**Comment by MuhammedZakir** [comment]
> [...] but maybe I can ask around on the Karabiner github or something.

I guess we'll have to do that.

> I needed to go through all of those steps just to get KMonad to run, and I followed the Mac installation instructions to the letter. I think, however, I'm basically just happy that I can see it compiles correctly without intending to tinker with Mac alot myself. It could very well be an issue with running in a virtualbox, since Mac already makes that very annoying, [...]

It was working fine the last time I tried, but I wanted to someone else to try it out, just in case. :-)

About this whole permission issues and stuff: I did get the exact issue you got, but I don't remember what I did to fix it. That's why, I mentioned all three possible fixes I can think of. Anyway, I will undo whatever I have done, and try to "install" KMonad again to verify the installation instructions.

> [...] and I think, technically in breach of the user-agreement...

It is a breach of user-agreement.

---

> How should we proceed from here? Shall I have a go at merging this PR in or are there any remaining issues that you are aware of on the Mac side of things?

There is no "issue" per se. However, I wanted to do one test\* regarding ignoring `Keyboard_Reserved` events. I will do that tomorrow. Other than that, everything is good.

\* Although technically there shouldn't be any problem, I wanted to make sure it's working as intended.

> After we merge this in I have some serious tinkering to do with the Windows stuff, that code isn't even finished yet, but I have had success in the past developing through VirtualBox.

Feel free to ping me if you need any help.

**Comment by MuhammedZakir** [comment]
Don't merge this. I am getting a weird behavior; when tapping `x` for the time, I am getting `P(0x7, 0x1) Rx P(0x7, 0x1) P(0x7, 0x1) Px P(0x7,0x1) Px`. For the subsequent taps of `x`, I am getting `P(0x7, 0x1) Rx P(0x7, 0x1) Rx P(0x7, 0x1) Px P(0x7,0x1) Px`. I am not sure what is the problem here. I am pretty sure I didn't get this the last time I tested. I don't know what went wrong. :-( Need sometime to debug this.

Btw, `(0x7, 0x1)` is `ErrorRollOver`.

**Comment by david-janssen** [comment]
No worries, I have some work to do on the Windows code anyways. I'm very happy leaving OSX stuff in your hands.

**Comment by MuhammedZakir** [comment]
I fixed it. However, I am still getting events for `(0x7, 0x1)`, but it's not being send/"put". That is, if I press `x`, I get
(note: `ErrorRollOver` == `(0x7, 0x1)`)
```
Got: Release ErrorRollOver
Got: Press x
Running Hooks
Running Hooks
Put: Press x
Registering untimed hook: 1
Got: Release ErrorRollOver
Got: Release x
Running hooks
Registering untimed hook: 2
Running hooks
Put: Release x
```

I am confused:
1. Why is `ErrorRollOver` not being send/"put"?
2. Will `ErrorRollOver` release event cause problem for some buttons? Should we ignore it?

---

> > I don't have an ISO keyboard, so this is just my guess: I think the key between lshift and z is (0x7, 0x64) - NonUS Backslash and the compose key is probably (0x7, 0x65) - Application. If you (or someone else) have ISO keyboard, then you can get the correct keycode using Karabiner-EventViewer.
> 
> Those both sound very plausible. I don't have an ISO keyboard I can plug into anything

From https://gist.github.com/MightyPork/6da26e382a7ad91b5496ee55fdc73db2#file-usb_hid_keys-h-L142:
```c
#define KEY_102ND 0x64 // Keyboard Non-US \ and |
#define KEY_COMPOSE 0x65 // Keyboard Application
```

**Comment by david-janssen** [comment]
> 1. Why is `ErrorRollOver` not being send/"put"?
> 2. Will `ErrorRollOver` release event cause problem for some buttons? Should we ignore it?

In regards to `1.`, I'm not entirely sure, since the way that MacOS codes its event-stream is still a mystery to me. However, in regards to `2.`, I think a very firm *yes* here, since `ErrorRollOver` is an event meant to indicate something semantically along the lines of: "Too many keys were pressed at the same time, so I don't really know what is pressed". The only thing we could sanely do with that is simply nothing. 

In the end, the input stream that should arrive at KMonad haskell code is a series of press-release alternations. e.g. `Pa Pb Ra Pc Rb Rc`. I think simply on the merit of the keymap coding we automatically ignore repeated presses or releases, but it might be good to even make it explicit.

> From https://gist.github.com/MightyPork/6da26e382a7ad91b5496ee55fdc73db2#file-usb_hid_keys-h-L142:
> 
> ```c
> #define KEY_102ND 0x64 // Keyboard Non-US \ and |
> #define KEY_COMPOSE 0x65 // Keyboard Application
> ```

Ahhh, great. You confirmed it :-). I'll uncomment the aliases in the `Parser`, did you add the names already to the mac keyboard definition? I suggest we stick with "102d" and "cmps" as the primary internal names for these keys, but we can add aliases like "nonUS" and "app" to the Parser.

**Comment by david-janssen** [comment]
Just to keep you in the loop, I updated all the keycode names for Windows, so now we are compiling fine under windows and a full 100% keyboard works without issue. The last thing I have to fix for windows is key-repeat, but after that both Linux and Windows will be ready. I'm awaiting your final "OK" to merge in the mac stuff, and after that I'll take some time to carefully rebase this ontop of master. Then a bit of testing, and maybe we can roll this out by the end of the week :-)

**Comment by MuhammedZakir** [comment]
In the new commit, `ErrorRollOver` keycode is also filtered. I have also changed keycode names for `non-US \` and `Application` in Mac to `102d` and `cmps`. And, fixed the aliases.

---

> The last thing I have to fix for windows is key-repeat, but after that both Linux and Windows will be ready.

Currently, there in no key-repeat if KMonad is used in Linux virtual console. IIRC, key-repeat is not baked into Windows, and if that's the case, it would be nice if the key-repeat feature is made generic, so that we could use it in Linux, and if need, in Mac too.

---

There is a bug in Mac related to modifier keys when more than 1 KMonad app is run at the same time. Because what needs to be fixed is mainly in C++ side, for now, we can merge this and I will open a new PR when I find a solution for it.

**Comment by MuhammedZakir** [comment]
Btw, do you need me to squash unnecessary commits and rebase them to get a clean working tree?

**Comment by david-janssen** [comment]
> Btw, do you need me to squash unnecessary commits and rebase them to get a clean working tree?

Nah, I was planning on doing that when we merge `keycode-refactor` onto master.

**Comment by david-janssen** [comment]
Wahay!
